kubernetes/controller/internal/resolution/workers.go

```diff
@@ -0,0 +1,274 @@
+	// Check cache first (fast path)
+	if cached, ok := wp.cache.Get(key); ok {
+		CacheHitCounterTotal.WithLabelValues(opts.Component, opts.Version).Inc()
+		if cached.err != nil {
+			return nil, cached.err
+		}
+		return cached.result, nil
+	}
+
+	CacheMissCounterTotal.WithLabelValues(opts.Component, opts.Version).Inc()
+
+	workItem := &WorkItem{
+		Context: ctx,
+		Key:     key,
+		Opts:    opts,
+		Cfg:     cfg,
+	}
+
+	select {
+	// Try to enqueue the request. If it fails, we return a queue full error so the reconciler can try again later.
+	case wp.workQueue <- workItem:
+		QueueSizeGauge.Set(float64(len(wp.workQueue)))
+		wp.logger.V(1).Info("enqueued resolution request", "component", opts.Component, "version", opts.Version)
+		return nil, ErrResolutionInProgress
+	default:
+		return nil, fmt.Errorf("lookup queue is full, cannot enqueue request for %s:%s", opts.Component, opts.Version)
+	}
+}
+
+// startWorker creates a worker's result channel and starts the worker in a goroutine.
+// Returns the channel that the worker will send results to.
+func (wp *WorkerPool) startWorker(ctx context.Context, id int) chan *Result {
+	resultChan := make(chan *Result)
+	logger := wp.logger.WithValues("worker", id)
+
+	go func() {
+		defer close(resultChan)
+		defer logger.V(1).Info("worker stopped", "id", id)
+
+		for {
+			select {
+			case <-ctx.Done():
+				logger.V(1).Info("worker stopped due to context cancellation")
+				return
+			case item, ok := <-wp.workQueue:
+				if !ok {
+					logger.V(1).Info("work queue closed, worker exiting")
+					return
+				}
+
+				QueueSizeGauge.Set(float64(len(wp.workQueue)))
+
+				logger.V(1).Info("processing work item", "key", item.Key)
+
+				start := time.Now()
+				result, err := wp.resolve(item.Context, &item.Opts, item.Cfg)
+				duration := time.Since(start).Seconds()
+
+				// Track metrics
+				ResolutionDurationHistogram.WithLabelValues(item.Opts.Component, item.Opts.Version).Observe(duration)
+
+				if err != nil {
+					logger.Error(err, "failed to process work item",
+						"component", item.Opts.Component,
+						"version", item.Opts.Version,
+						"duration", duration)
+				} else {
+					logger.V(1).Info("processed work item",
+						"component", item.Opts.Component,
+						"version", item.Opts.Version,
+						"duration", duration)
+				}
+
+				// Send result to worker's dedicated result channel
+				resultChan <- &Result{
+					key:      item.Key,
+					result:   result,
+					err:      err,
+					createAt: time.Now(),
+				}
+			}
+		}
+	}()
+
+	return resultChan
+}
+
+// resultCollector processes results from multiple worker channels and updates the cache.
+func (wp *WorkerPool) resultCollector(ctx context.Context, workerChannels []chan *Result) {
+	logger := wp.logger.WithValues("component", "result-collector")
+	logger.V(1).Info("result collector started", "workerCount", len(workerChannels))
+
+	// Merge all worker channels into a single channel (fan-in)
+	mergedResults := make(chan *Result)
```

Comment:
instead of using a waitgroup and a go routine to add into mergedResults, couldnt you simply buffer the channel with size len(workerChannels) here?

Reply:
I think Im surprised that we even have multiple worker channels tbh. why cant this be shared?